main/network.c

```diff
@@ -1274,91 +1274,51 @@ PHPAPI int php_poll2(php_pollfd *ufds, unsigned int nfds, int timeout)
 }
 #endif
 
-#if defined(HAVE_GETHOSTBYNAME_R)
-#ifdef HAVE_FUNC_GETHOSTBYNAME_R_6
-struct hostent * gethostname_re (const char *host,struct hostent *hostbuf,char **tmphstbuf,size_t *hstbuflen)
-{
-	struct hostent *hp;
-	int herr,res;
-
-	if (*hstbuflen == 0) {
-		*hstbuflen = 1024;
-		*tmphstbuf = (char *)malloc (*hstbuflen);
-	}
-
-	while (( res =
-		gethostbyname_r(host,hostbuf,*tmphstbuf,*hstbuflen,&hp,&herr))
-		&& (errno == ERANGE)) {
-		/* Enlarge the buffer. */
-		*hstbuflen *= 2;
-		*tmphstbuf = (char *)realloc (*tmphstbuf,*hstbuflen);
-	}
-
-	if (res != SUCCESS) {
-		return NULL;
-	}
-
-	return hp;
-}
-#endif
-#ifdef HAVE_FUNC_GETHOSTBYNAME_R_5
-struct hostent * gethostname_re (const char *host,struct hostent *hostbuf,char **tmphstbuf,size_t *hstbuflen)
-{
-	struct hostent *hp;
-	int herr;
-
-	if (*hstbuflen == 0) {
-		*hstbuflen = 1024;
-		*tmphstbuf = (char *)malloc (*hstbuflen);
-	}
-
-	while ((NULL == ( hp =
-		gethostbyname_r(host,hostbuf,*tmphstbuf,*hstbuflen,&herr)))
-		&& (errno == ERANGE)) {
-		/* Enlarge the buffer. */
-		*hstbuflen *= 2;
-		*tmphstbuf = (char *)realloc (*tmphstbuf,*hstbuflen);
-	}
-	return hp;
-}
-#endif
-#ifdef HAVE_FUNC_GETHOSTBYNAME_R_3
-struct hostent * gethostname_re (const char *host,struct hostent *hostbuf,char **tmphstbuf,size_t *hstbuflen)
+#if HAVE_GETADDRINFO
+struct hostent* gethostname_re (const char *host)
 {
-	if (*hstbuflen == 0) {
-		*hstbuflen = sizeof(struct hostent_data);
-		*tmphstbuf = (char *)malloc (*hstbuflen);
-	} else {
-		if (*hstbuflen < sizeof(struct hostent_data)) {
-			*hstbuflen = sizeof(struct hostent_data);
-			*tmphstbuf = (char *)realloc(*tmphstbuf, *hstbuflen);
-		}
-	}
-	memset((void *)(*tmphstbuf),0,*hstbuflen);
-
-	if (SUCCESS != gethostbyname_r(host,hostbuf,(struct hostent_data *)*tmphstbuf)) {
-		return NULL;
-	}
-
-	return hostbuf;
+    struct hostent *hp;
+    struct addrinfo hints, *result, *rp;
+    unsigned int nresults = 0;
+    char* nullp = NULL;
+	int i;
+
+    memset(&hints, 0, sizeof(struct addrinfo));
+    hints.ai_family = PF_INET;
+    hints.ai_socktype = SOCK_DGRAM;
+
+    if (getaddrinfo(host, NULL, &hints, &result) != SUCCESS) {
+        return NULL;
+    }
+
+    for (rp = result; rp != NULL; rp = rp->ai_next)
+        nresults++;
+
+    hp = emalloc(sizeof(struct hostent));
+    hp->h_name = strdup(host);
```

Comment:
```suggestion
    hp->h_name = strndup(host, 255);
```
I suggest we guard against a buffer overflow attack here. 
255 bytes will be enough to fit any valid hostname per RFC 1035.